### ISSUE TITLE
Permissions

### ISSUE THREAD

**Issue by renatomefidf** [RobLoach/docker-composer#35]
Hello,

If I use composer without changing any settings my vendor will get root:root 755 permissions.
If I use my own UID for user and/or group it will fail on writing into the cache.

Am I doing something wrong or is this the expected behavious?

**Comment by RobLoach** [comment]
Could you post your workflow to get the permission error?

**Comment by fesor** [comment]
@RobLoach here is my workaround:

```
FROM composer/composer:php7

RUN git config --system user.name "Docker" && \
    git config --system user.email "docker <at> localhost"

ENV COMPOSER_HOME="/composer"

```

And here how I use it:

```
docker run --rm -it \
     --user $(id -u) \
     -v $(pwd):/app \
     -v ~/.composer:/composer \
     fesor/composer $@
```

This solves:

 - issues of permissions, since container is running from current user
 - git issues, which fails to clone repo is no such user exists in `/etc/passwd`
 - make use of cache and composer configs, since it uses user's configs and cache via volumes

**Comment by renatomefidf** [comment]
@RobLoach its exactly as @fesor showed, I'm using your image as a volatile container with --rm option.
Still its a lot of "hack" to use this solution, mainly because you have to set git user.

Have you see this approach: https://github.com/eXistenZNL/Docker-Composer ?

It seems pretty effect for this kind of usage!

Thank you

**Comment by RobLoach** [comment]
That sounds like a great solution. So, in summary, we should:

1. Add the following to the Dockerfile base:
  ```
  RUN git config --system user.name "Docker" && \
  git config --system user.email "docker <at> localhost"
  ```

2. Update usage documentation to:
  ```
  docker run --rm -it \
     --user $(id -u) \
     -v $(pwd):/app \
     -v ~/.composer:/composer \
     composer/composer $@
  ```

3. Switch COMPOSER_HOME to `/composer`

**Comment by fesor** [comment]
@RobLoach yep, the only case that could go wrong is handling of private repositories. Will to check this.

**Comment by dagomar** [comment]
I'm using this solution, will this be part of the official build??

**Comment by roverwolf** [comment]
The second $(id -u) in the previous comment should probably be $(id -g) to get the group id.

In order to get rid of the `fatal: unable to look up current user in the passwd file: no such user` error from git when it needs to clone a repo I've found I need to add:
```
    -v /etc/passwd:/etc/passwd
```

**Comment by fesor** [comment]
@roverwolf, I solve this just by adding user in got config:

```
RUN git config --system user.name "Docker" \
   && git config --system user.email "docker <at> localhost"
```

This solves this issue

**Comment by roverwolf** [comment]
@fesor, I tried running the git config commands in my Dockerfile, but it did not seem to work.  I'll try again as I may have misconfigured something along the way.

Bind mounting `/etc/passwd` (or a subset of it) into the container is another option for anyone running into the issue if setting the user name and email doesn't work for them.

**Comment by llernestal** [comment]
To get the container working with my private repositories I used this:

```bash
docker run -it --rm \
--user $(id -u):$(id -g) \
-v $(pwd):/app \
-v ~/.composer:/composer \
-v /etc/passwd:/etc/passwd \
-v ~/.ssh:$HOME/.ssh \
composer/composer $@
```

**Comment by toby-griffiths** [comment]
I'm using the following command, which works OK…

```
docker run --rm --user $(id -u):$(id -g) -v $(pwd):/app composer/composer install
```

… but I get the warning… 

```
Cannot create cache directory /composer/cache/repo/https---packagist.org/, or directory is not writable. Proceeding without cache
Cannot create cache directory /composer/cache/files/, or directory is not writable. Proceeding without cache
```

Any ideas how to get around this?

**Comment by llernestal** [comment]
@toby-griffiths, Try adding `-v ~/.composer:/composer` to your command

**Comment by toby-griffiths** [comment]
@llernestal Thanks :o)
